Move the SmartUnion's active member enforcement into method body

nemanja-boric-sociomantic opened the pull request:
The enforcement which was used to confirm that the active member
has been accessed was omitted in the `-release` build. Now it's moved
in the method's body and it still active, no matter if compiled with
`-release`.

Fixes #96

nemanja-boric-sociomantic committed:
Move the SmartUnion's active member enforcement into method body

The enforcement which was used to confirm that the active member
has been accessed was omitted in the `-release` build. Now it's moved
in the method's body and it still active, no matter if compiled with
`-release`.

Fixes #96

mihails-strasuns-sociomantic commented:
`enforce` or `verify`? Is the intent of smart union that one shall never call `get` with wrong type?

nemanja-boric-sociomantic commented:
I think so, I'll use verify.

nemanja-boric-sociomantic committed:
Change enforce in SmartUnion.get to verify

The intention is that SmartUnion's wrong type is never accessed and
that this should be enforced by the programmer, not by the environment,
so this should actually be _verify_.

nemanja-boric-sociomantic committed:
Test ocean with -release

nemanja-boric-sociomantic commented:
Updated. I've also added the matrix item with `-release`, let's see if that work.

nemanja-boric-sociomantic commented:
Ok, passing `DFLAGS=-release` to travis works:

https://travis-ci.org/sociomantic-tsunami/ocean/jobs/272091389

gavin-norman-sociomantic commented on src/ocean/core/SmartUnion.d:
```diff
@@ -503,9 +503,9 @@ private template Methods ( U, uint i )
     const type = "typeof(" ~ member_access ~ ")";
 
     const get = type ~ ' ' ~  member ~ "() "
-        ~ "in { enforce(_.active == _.active." ~ member ~ ", "
```
Oof sneaky!
